accord-core/src/main/java/accord/coordinate/CheckOn.java

```diff
@@ -0,0 +1,236 @@
+    /**
+     * The epoch until which we want to persist any response for locally
+     */
+    final ReplicationPhase sufficient;
+    final long untilLocalEpoch;
+    final AbstractRoute route;
+
+    CheckOn(Node node, ReplicationPhase sufficient, TxnId txnId, AbstractRoute route, long srcEpoch, long untilLocalEpoch, BiConsumer<? super CheckStatusOkFull, Throwable> callback)
+    {
+        this(node, sufficient, txnId, route, route.with(route.homeKey), srcEpoch, untilLocalEpoch, callback);
+    }
+
+    CheckOn(Node node, ReplicationPhase sufficient, TxnId txnId, AbstractRoute route, RoutingKeys routeWithHomeKey, long srcEpoch, long untilLocalEpoch, BiConsumer<? super CheckStatusOkFull, Throwable> callback)
+    {
+        // TODO (soon): restore behaviour of only collecting info if e.g. Committed or Executed
+        super(node, txnId, routeWithHomeKey, srcEpoch, IncludeInfo.All);
+        Preconditions.checkArgument(routeWithHomeKey.contains(route.homeKey));
+        this.sufficient = sufficient;
+        this.route = route;
+        this.callback = callback;
+        this.untilLocalEpoch = untilLocalEpoch;
+    }
+
+    // TODO: many callers only need to consult precisely executeAt.epoch remotely
+    public static CheckOn checkOn(ReplicationPhase sufficientStatus, Node node, TxnId txnId, AbstractRoute route, long srcEpoch, long untilLocalEpoch, BiConsumer<? super CheckStatusOkFull, Throwable> callback)
+    {
+        CheckOn checkOnCommitted = new CheckOn(node, sufficientStatus, txnId, route, srcEpoch, untilLocalEpoch, callback);
+        checkOnCommitted.start();
+        return checkOnCommitted;
+    }
+
+    public static CheckOn checkOnUncommitted(Node node, TxnId txnId, AbstractRoute route, long untilLocalEpoch, BiConsumer<? super CheckStatusOkFull, Throwable> callback)
+    {
+        return checkOn(Commit, node, txnId, route, txnId.epoch, untilLocalEpoch, callback);
+    }
+
+    protected AbstractRoute route()
+    {
+        return route;
+    }
+
+    @Override
+    protected boolean isSufficient(Id from, CheckStatusOk ok)
+    {
+        return ((CheckStatusOkFull)ok).sufficientFor(route()).compareTo(sufficient) >= 0;
+    }
+
+    @Override
+    protected void onDone(Done done, Throwable failure)
+    {
+        if (failure != null)
+        {
+            callback.accept(null, failure);
+        }
+        else if (merged == null || merged.status == NotWitnessed)
+        {
+            callback.accept(CheckStatusOkFull.NOT_WITNESSED, null);
+        }
+        else
+        {
+            new OnDone();
+        }
+    }
+
+    class OnDone implements MapReduceConsume<SafeCommandStore, Void>
+    {
+        final AbstractRoute maxRoute;
+        final RoutingKey progressKey;
+        final CheckStatusOkFull full;
+        final ReplicationPhase sufficientTo;
+        final PartialTxn partialTxn;
+        final PartialDeps partialDeps;
+
+        public OnDone()
```

Comment:
Yep. It's uglier in this case to artificially introduce a separate method to trigger the `mapReduceConsume`.

Reply:
I forgot why this is a bad idea and had to read up, but basically leaking this from a constructor in a multi-threaded context is something you aren't supposed to do since you are publishing a reference to a partially constructed object.

Since publication is the last line of the constructor and there is probably something in publishing this to another thread that adds the necessary barriers it might be safe. It also might be the JMM says it is unsafe.

Generally I don't attempt to publish an object to other threads from the constructor.

Reply:
So yeah, we probably don’t get the guarantees of final but are safe publishing anyway via volatile semantics. That said, I will refactor anyway to avoid future head scratching.

Reply:
This should be resolved now.